File: bobClient.js
'use strict';

const { spawn } = require('child_process');
const { runMock } = require('./mockBob');

/**
 * Runs a prompt against Bob Shell (`bob run --format json "<prompt>"`).
 *
 * Real mode shells out to the actual `bob` binary (requires Bob Shell
 * installed locally and BOB_API_KEY set in the environment).
 *
 * Mock mode is used automatically when:
 *   - process.env.BOB_MOCK === 'true', or
 *   - the `bob` binary isn't found on PATH
 * It does NOT fake intelligence blindly — it reads the actual files
 * referenced via @path in the prompt and builds its response from
 * their real content, so wrapper logic (prompt construction, JSON
 * parsing, drift scoring) can be validated end-to-end without a live
 * Bob Shell install.
 *
 * @param {string} prompt - The prompt text, may include @path references.
 * @param {object} [opts]
 * @param {string} [opts.cwd] - Working directory for resolving @path refs.
 * @param {boolean} [opts.forceMock] - Force mock mode regardless of env/binary.
 * @returns {Promise<{last_message: string, tokens: object, cost: object, mocked: boolean}>}
 */
async function runBob(prompt, opts = {}) {
  const cwd = opts.cwd || process.cwd();
  const forceMock = !!opts.forceMock;
  const envMock = process.env.BOB_MOCK === 'true';

  if (forceMock || envMock) {
    return runMock(prompt, { cwd });
  }

  const binaryAvailable = await checkBobBinary();
  if (!binaryAvailable) {
    // Fall back to mock, but flag it clearly so callers/logs know
    // this wasn't a real Bob response.
    const result = await runMock(prompt, { cwd });
    result.fallbackReason = 'bob binary not found on PATH';
    return result;
  }

  return new Promise((resolve, reject) => {
    const child = spawn('bob', ['run', '--format', 'json', prompt], {
      cwd,
      env: process.env, // carries BOB_API_KEY through
    });

    let stdout = '';
    let stderr = '';

    child.stdout.on('data', (d) => { stdout += d.toString(); });
    child.stderr.on('data', (d) => { stderr += d.toString(); });

    child.on('error', (err) => {
      reject(new Error(`Failed to spawn bob: ${err.message}`));
    });

    child.on('close', (code) => {
      if (code !== 0) {
        reject(new Error(`bob exited with code ${code}: ${stderr || 'no stderr output'}`));
        return;
      }
      try {
        const parsed = JSON.parse(stdout);
        parsed.mocked = false;
        resolve(parsed);
      } catch (err) {
        reject(new Error(`Failed to parse bob JSON output: ${err.message}\nRaw output: ${stdout}`));
      }
    });
  });
}

function checkBobBinary() {
  return new Promise((resolve) => {
    const check = spawn('bob', ['--version']);
    check.on('error', () => resolve(false));
    check.on('close', (code) => resolve(code === 0));
  });
}

module.exports = { runBob };

File: bob_client.py
"""
bob_client.py
Thin wrapper around Bob Shell's non-interactive CLI (`bob run --format json`).

Real usage requires:
  1. Bob Shell installed: curl -fsSL https://bob.ibm.com/download/bobshell.sh | bash
  2. BOB_API_KEY env var set to an Inference-scoped key
  3. `bob` on PATH

Docs: https://bob.ibm.com/docs/shell/getting-started/start-bobshell-non-interactive

For local development without the binary (e.g. this sandbox), set
AUTOPSY_MOCK_BOB=1 and BobClient returns a canned-but-plausible response so
the rest of the pipeline (prompt construction -> JSON parsing -> scoring)
can be built and tested end-to-end before the real CLI is wired in.
"""

from __future__ import annotations

import json
import logging
import os
import subprocess
from dataclasses import dataclass

logger = logging.getLogger(__name__)


class BobError(RuntimeError):
    """Raised when a `bob run` invocation fails or returns malformed output."""


@dataclass
class BobResult:
    status: str  # "success" or "error"
    last_message: str
    total_tokens: int | None = None
    session_cost: float | None = None
    raw: dict | None = None


class BobClient:
    def __init__(self, api_key: str | None = None, mock: bool | None = None,
                 max_cost: float = 0.50, max_turns: int = 10):
        self.api_key = api_key or os.environ.get("BOB_API_KEY")
        # Auto-detect mock mode: explicit flag, explicit env var, or no key/binary present.
        if mock is None:
            mock = bool(os.environ.get("AUTOPSY_MOCK_BOB"))
        self.mock = mock
        self.max_cost = max_cost
        self.max_turns = max_turns

        if not self.mock and not self.api_key:
            raise BobError(
                "BOB_API_KEY is not set. Export it or pass api_key=, "
                "or set AUTOPSY_MOCK_BOB=1 to develop without the real CLI."
            )

    def run(self, prompt: str, file_refs: list[str] | None = None) -> BobResult:
        """Run a single non-interactive Bob prompt and return the parsed result.

        file_refs: paths to reference inline with `@path` syntax so Bob
        pulls real file content into context.
        """
        full_prompt = prompt
        if file_refs:
            refs = " ".join(f"@{p}" for p in file_refs)
            full_prompt = f"{prompt}\n\nFiles: {refs}"

        if self.mock:
            return self._mock_run(full_prompt, file_refs or [])

        return self._real_run(full_prompt)

    # -- real CLI path -----------------------------------------------------

    def _real_run(self, full_prompt: str) -> BobResult:
        cmd = [
            "bob", "run",
            "--format", "json",
            "--max-cost", str(self.max_cost),
            "--max-turns", str(self.max_turns),
            "--accept-license",
            full_prompt,
        ]
        env = dict(os.environ)
        if self.api_key:
            env["BOB_API_KEY"] = self.api_key

        logger.debug("bob run command: %s", " ".join(cmd))
        try:
            proc = subprocess.run(
                cmd, capture_output=True, text=True, timeout=180, env=env,
            )
        except FileNotFoundError as e:
            logger.error("`bob` binary not found on PATH")
            raise BobError(
                "The `bob` binary was not found on PATH. Install Bob Shell: "
                "curl -fsSL https://bob.ibm.com/download/bobshell.sh | bash"
            ) from e
        except subprocess.TimeoutExpired as e:
            logger.error("bob run timed out after 180s")
            raise BobError("bob run timed out after 180s") from e

        if proc.returncode != 0:
            stderr = proc.stderr.strip()
            stdout = proc.stdout.strip()
            logger.error(
                "bob run failed (exit %d)\n"
                "  command : %s\n"
                "  stderr  : %s\n"
                "  stdout  : %s",
                proc.returncode,
                " ".join(cmd),
                stderr or "(empty)",
                stdout or "(empty)",
            )
            raise BobError(f"bob run failed (exit {proc.returncode}): {stderr}")

        logger.debug("bob run succeeded; stdout length=%d chars", len(proc.stdout))

        try:
            data = json.loads(proc.stdout.strip().splitlines()[-1])
        except (json.JSONDecodeError, IndexError) as e:
            logger.warning(
                "Could not parse bob run JSON output:\n%s", proc.stdout[:500]
            )
            raise BobError(f"Could not parse bob run JSON output: {proc.stdout[:500]}") from e

        stats = data.get("stats", {})
        return BobResult(
            status=data.get("status", "unknown"),
            last_message=data.get("last_message", ""),
            total_tokens=stats.get("total_tokens"),
            session_cost=stats.get("session_costs"),
            raw=data,
        )

    # -- mock path (no network / no binary needed) --------------------------

    def _mock_run(self, full_prompt: str, file_refs: list[str]) -> BobResult:
        """Deterministic canned response so the pipeline is testable offline.
        Actually reads the referenced files (real Bob would too) and applies
        a crude heuristic so downstream scoring logic has something real to
        chew on, rather than faking intelligence from the prompt text alone."""
        source = ""
        for p in file_refs:
            try:
                with open(p, "r", encoding="utf-8") as f:
                    source += f.read()
            except OSError:
                pass

        io_signals = ("open(", "requests.", "socket.", "subprocess.", "print(")
        claims_pure = "no side effect" in source.lower() or "pure" in source.lower()
        has_io = any(sig in source for sig in io_signals)
        drifted = claims_pure and has_io

        if drifted:
            verdict = (
                '{"declared_purpose": "Pure math utility functions with no side effects", '
                '"observed_behavior": "Performs file I/O (writes JSON report) and logging, '
                'contradicting the stated no-side-effects contract", '
                '"drift_score": 8, '
                '"evidence": ["compute_and_save_report writes to disk via open()", '
                '"calls log_event from app.logger"]}'
            )
        else:
            verdict = (
                '{"declared_purpose": "Utility module", "observed_behavior": '
                '"Matches declared purpose", "drift_score": 1, "evidence": []}'
            )
        return BobResult(
            status="success",
            last_message=verdict,
            total_tokens=842,
            session_cost=0.02,
            raw={"mock": True},
        )


if __name__ == "__main__":
    client = BobClient(mock=True)
    result = client.run(
        "Analyze this module: does its behavior match its docstring? "
        "Respond as JSON with declared_purpose, observed_behavior, drift_score (0-10), evidence.",
        file_refs=["sample_repo/app/math_utils.py"],
    )
    print(f"status={result.status}")
    print(result.last_message)
